src/storage/mvcc/reader.rs

```diff
@@ -417,4 +418,236 @@ impl<'a> MvccReader<'a> {
             keys.push(key);
         }
     }
+
+    // Returns true if it needs gc.
+    // This is for optimization purpose, does not mean to be accurate.
+    pub fn need_gc(&self, safe_point: u64) -> bool {
+        let mut opts = GetPropertiesOptions::default();
+        opts.max_ts = Some(safe_point);
+        let props = match self.snapshot.get_properties_cf(CF_WRITE, &opts) {
+            Ok(v) => v,
+            Err(_) => return true,
+        };
+        // We don't need gc if:
+        // 1. min_ts > safe_point:
+        //    No data older than safe_point to gc.
+        if props.min_ts > safe_point {
+            return false;
+        }
+        // 2. num_rows == num_versions && num_puts == num_versions:
+        //    No rows have more than one versions, and the version is effective.
+        //    Notice: Since the properties are file-based, it can be false positive.
+        //    For example, if multiple files have a different version of the same row.
+        !(props.num_rows == props.num_versions && props.num_puts == props.num_versions)
```

Comment:
2) is not covered by 1), for example we execute sql like `delete * from t...`, that will add a mvcc_del for each row we have deleted, the old version mvcc_put may be dropped after gc  finished, but the fresh mvcc_del will left, these mvcc_del also need gc.
3) is not covered by 1) too, you have finger it out above, I think 100 is enough.